build.py: use 'pathlib' module and f-Strings

gdoffe opened the pull request:
This commit does not introduce functionnal changes.
It makes the code more compliant with Python 3 standards by
using 'pathlib' module [1] and f-strings format [2].

[1]: https://docs.python.org/fr/3/library/pathlib.html
[2]: https://docs.python.org/fr/3/tutorial/inputoutput.html#formatted-string-literals

Signed-off-by: Gilles DOFFE <g.doffe@gmail.com>

gdoffe committed:
build.py: use 'pathlib' module and f-Strings

This commit does not introduce functionnal changes.
It makes the code more compliant with Python 3 standards by
using 'pathlib' module [1] and f-strings format [2].

[1]: https://docs.python.org/fr/3/library/pathlib.html
[2]: https://docs.python.org/fr/3/tutorial/inputoutput.html#formatted-string-literals

Signed-off-by: Gilles DOFFE <g.doffe@gmail.com>

ecourtois commented on build.py:
```diff
@@ -2,6 +2,7 @@
 
 import glob
 import os
+import pathlib
```
```python
from pathlib import Path
```

Then use `Path` instead of `pathlib.Path`.

gdoffe replied:
done

ecourtois commented on build.py:
```diff
@@ -28,14 +29,17 @@ def __init__(self):
-            "./layers/**/oe-init-build-env", recursive=True)
+        oe_init_build_env = list(
+                pathlib.Path().glob("layers/**/oe-init-build-env")
+                )
```
I would prefer:

```python
self.layer_path = Path("layers")
oe_init_build_env = list(self.layers_path.glob("**/oe-init-build-env"))
```

gdoffe replied:
done

ecourtois commented on build.py:
```diff
@@ -28,14 +29,17 @@ def __init__(self):
-            + str(oe_init_build_env)
-        self.cmd_prefix = ". " + os.path.abspath(oe_init_build_env[0]) + " ; "
+            "Only one 'oe-init-build-env' can be used, several found: {}\n".format(
+            oe_init_build_env)
```
By using f-string, I mean formating like this:

```python
f"Only one 'oe-init-build-env' can be used, several found: {oe_init_build_env}\n"
```

Same for other strings.

gdoffe replied:
done

ecourtois commented on build.py:
```diff
@@ -28,14 +29,17 @@ def __init__(self):
+            "Only one 'oe-init-build-env' can be used, several found: {}\n".format(
+            oe_init_build_env)
+        self.cmd_prefix = ". {};".format(
+                (pathlib.Path(oe_init_build_env[0]).resolve())
```
This is already a `Path`, no need to cast it again.

gdoffe replied:
done

ecourtois commented on build.py:
```diff
@@ -45,19 +49,17 @@ def _add_layers(self):
-            "./layers/**/meta-*/conf/layer.conf", recursive=True)
-        layers = list(map(lambda x: os.path.abspath(
-            x.replace('/conf/layer.conf', '')), layers))
+        layers = pathlib.Path().glob("./layers/**/meta-*/conf/layer.conf")
```
Same as before, and you can now reuse `self.layers_path`.

gdoffe replied:
done

ecourtois commented on build.py:
```diff
@@ -28,14 +29,17 @@ def __init__(self):
+                (pathlib.Path(oe_init_build_env[0]).resolve())
+                )
 
         self._add_layers()
```
You should not call functions that do real stuff in the constructor.
Please call it after the constructor.

gdoffe replied:
done

ecourtois reviewed (changes requested):
You also missing my comment about `subprocess` usage ;-)
Using `subprocess.Popen` should be avoided when possible, it is the low level API.
Prefer using [`subprocess.run`](https://docs.python.org/3.8/library/subprocess.html#subprocess.run) .

ecourtois reviewed (changes requested):
`import glob` is unused.

gdoffe commented:
> You also missing my comment about `subprocess` usage ;-) Using `subprocess.Popen` should be avoided when possible, it is the low level API. Prefer using [`subprocess.run`](https://docs.python.org/3.8/library/subprocess.html#subprocess.run) .

Not that was wanted. ;)
With `subprocess.Popen` you can wait for the child process to end, not with `subprocess.run`.
The documentation says that `subprocess.run` `Wait for command to complete, then return a CompletedProcess instance.`
https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module

However `subprocess.run` returns nearly immediatly without waiting. I do not know why. So I used `Popen` and did the `wait()` manually.

gdoffe committed:
build.py: refactor run() method

This commit does not add functionnal changes.
Use subprocess.run() instead of subprocess.Popen() as the Python
documentation recommands:
  The recommended approach to invoking subprocesses is to use
  the run() function for all use cases it can handle. For more
  advanced use cases, the underlying Popen interface can be used
  directly. [1]

[1]: https://docs.python.org/3/library/subprocess.html#subprocess.Popen

Signed-off-by: Gilles DOFFE <g.doffe@gmail.com>

gdoffe commented:
> > You also missing my comment about `subprocess` usage ;-) Using `subprocess.Popen` should be avoided when possible, it is the low level API. Prefer using [`subprocess.run`](https://docs.python.org/3.8/library/subprocess.html#subprocess.run) .
> 
> Not that was wanted. ;) With `subprocess.Popen` you can wait for the child process to end, not with `subprocess.run`. The documentation says that `subprocess.run` `Wait for command to complete, then return a CompletedProcess instance.` https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module
> 
> However `subprocess.run` returns nearly immediatly without waiting. I do not know why. So I used `Popen` and did the `wait()` manually.

done ;)

ecourtois reviewed (approved):
LGTM
